core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java

```diff
@@ -0,0 +1,98 @@
+ * in the list is the target service account. Intermediate delegated credentials are applied first
+ * from end of the list, working towards the beginning of the list. For example, if <code>
+ * delegates = Arrays.asList( "first@serviceaccount.com", "second@serviceaccount.com",
+ * "third@serviceaccount.com")
+ * </code>
+ *
+ * <p>The connector will start with the GoogleCredentials supplied by `source`. It will use those to
+ * impersonate "third@serviceaccount.com", then with a token for "third@serviceaccount.com",
+ * impersonate "second@serviceaccount.com", and then impersonate "first@serviceaccount.com".
+ * Finally, the connector will attempt to access the services as "first@serviceaccount.com".
+ */
+class ServiceAccountImpersonatingCredentialFactory implements CredentialFactory {
+
+  private final CredentialFactory source;
+  private final List<String> delegates;
+  private final String targetPrincipal;
+
+  ServiceAccountImpersonatingCredentialFactory(CredentialFactory source, List<String> delegates) {
+    if (delegates == null || delegates.isEmpty()) {
+      throw new IllegalArgumentException("delegates must not be empty");
+    }
+    this.source = source;
+
+    // The "delegates" property is built so that delegated credentials are applied first from end
+    // of the list, working towards the beginning of the list. However, the
+    // ImpersonatedCredentials.setDelegates() expects the list to be in the opposite order,
+    // so we have to reverse the list.
+    //
+    // From the ImpersonatedCredentials doc:
+    //
+    //    targetPrincipal – the service account to impersonate
+    //    delegates – the chained list of delegates required to grant the final access_token.
+    //      If set, the sequence of identities must have "Service Account Token Creator"
+    //      capability granted to the preceding identity. For example, if set to
+    //      [serviceAccountB, serviceAccountC], the sourceCredential must have the Token Creator
+    //      role on serviceAccountB. serviceAccountB must have the Token Creator on
+    //      serviceAccountC. Finally, C must have Token Creator on target_principal.
+    //      If unset, sourceCredential must have that role on targetPrincipal.
+    List<String> reversedDelegates = Lists.reverse(delegates);
+    this.targetPrincipal = reversedDelegates.get(reversedDelegates.size() - 1);
+    this.delegates = reversedDelegates.subList(0, reversedDelegates.size() - 1);
+  }
+
+  @Override
+  public HttpRequestInitializer create() {
+    GoogleCredentials credentials = getCredentials();
+    return new HttpCredentialsAdapter(credentials);
+  }
+
+  @Override
+  public GoogleCredentials getCredentials() {
+    GoogleCredentials credentials = source.getCredentials();
+
+    credentials =
+        ImpersonatedCredentials.newBuilder()
+            .setSourceCredentials(credentials)
+            .setTargetPrincipal(targetPrincipal)
+            .setDelegates(this.delegates)
```

Comment:
Should we conditionally set delegates? What happens if delegates is a list of size 1?